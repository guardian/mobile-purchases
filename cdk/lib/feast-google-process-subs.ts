import type { GuStackProps } from "@guardian/cdk/lib/constructs/core";
import { GuStack } from "@guardian/cdk/lib/constructs/core";
import { GuSnsLambdaExperimental } from "@guardian/cdk/lib/experimental/patterns";
import { Duration } from 'aws-cdk-lib';
import type { App } from "aws-cdk-lib";
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { Queue } from "aws-cdk-lib/aws-sqs";

export interface FeastSubscriptionsProps extends GuStackProps {
    stack: string;
    stage: string;
}


export class FeastGoogleProcessSubscriptions extends GuStack {
    constructor(scope: App, id: string, props: FeastSubscriptionsProps) {
        super(scope, id, props);

        const app = 'mobile-purchases-feast-google-process-subs';
        const nameWithStage = `${app}-${this.stage}`;

        const feastGoogleProcessSubsQueue = new Queue(this, `${nameWithStage}-sns-queue`, {
            queueName: `${nameWithStage}-queue`,
            retentionPeriod: Duration.days(14),
        });

        new GuSnsLambdaExperimental(this,`${ nameWithStage }-lambda`,{
            app: `${ nameWithStage }-lambda`,
            functionName: `${ nameWithStage }-lambda`,
            existingSnsTopic: { externalTopicName: feastGoogleProcessSubsQueue.queueName },
            runtime: Runtime.NODEJS_20_X,
            monitoringConfiguration: { noMonitoring: true },
            handler: 'index.handler',
            fileName: `${app}.zip`,
        });

        //permissions to read sns queue

        //permissions to access SMS and get google auth

        //permissions to make call to Google

        //permissions to read and write to database


    }

}