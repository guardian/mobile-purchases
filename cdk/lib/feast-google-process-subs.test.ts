import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import {FeastGoogleProcessSubscriptions } from "./feast-google-process-subs";

describe('The FeastAndroidSubscriptions stack', () => {
    it('matches the snapshot', () => {
        const app = new App();
        const codeStack = new FeastGoogleProcessSubscriptions(app, 'feast-subscriptions-CODE',{ stack: 'mobile', stage: 'CODE' });
        const prodStack = new FeastGoogleProcessSubscriptions(app, 'feast-subscriptions-PROD',{ stack: 'mobile', stage: 'PROD' });

        expect(Template.fromStack(codeStack).toJSON()).toMatchSnapshot();
        expect(Template.fromStack(prodStack).toJSON()).toMatchSnapshot();
    });
});