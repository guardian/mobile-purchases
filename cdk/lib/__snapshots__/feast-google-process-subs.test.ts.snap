// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The FeastAndroidSubscriptions stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuSnsLambdaExperimental",
    ],
    "gu:cdk:version": "TEST",
  },
  "Outputs": {
    "mobilepurchasesfeastgoogleprocesssubsCODElambdaTopicName3B8382CC": {
      "Value": {
        "Fn::GetAtt": [
          "mobilepurchasesfeastgoogleprocesssubsCODEsnsqueue5470C729",
          "QueueName",
        ],
      },
    },
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "mobilepurchasesfeastgoogleprocesssubsCODElambdaA1DA9075": {
      "DependsOn": [
        "mobilepurchasesfeastgoogleprocesssubsCODElambdaServiceRoleDefaultPolicy6C37C6FB",
        "mobilepurchasesfeastgoogleprocesssubsCODElambdaServiceRoleFA723C8D",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "mobile/CODE/mobile-purchases-feast-google-process-subs-CODE-lambda/mobile-purchases-feast-google-process-subs.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "mobile-purchases-feast-google-process-subs-CODE-lambda",
            "STACK": "mobile",
            "STAGE": "CODE",
          },
        },
        "FunctionName": "mobile-purchases-feast-google-process-subs-CODE-lambda",
        "Handler": "index.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "mobilepurchasesfeastgoogleprocesssubsCODElambdaServiceRoleFA723C8D",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "mobile-purchases-feast-google-process-subs-CODE-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "mobilepurchasesfeastgoogleprocesssubsCODElambdaAllowInvokefeastsubscriptionsCODEmobilepurchasesfeastgoogleprocesssubsCODElambdaSnsExistingIncomingEventsTopicD63F705E90C9ED09": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "mobilepurchasesfeastgoogleprocesssubsCODElambdaA1DA9075",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:sns:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::GetAtt": [
                  "mobilepurchasesfeastgoogleprocesssubsCODEsnsqueue5470C729",
                  "QueueName",
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "mobilepurchasesfeastgoogleprocesssubsCODElambdaServiceRoleDefaultPolicy6C37C6FB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/mobile/CODE/mobile-purchases-feast-google-process-subs-CODE-lambda/mobile-purchases-feast-google-process-subs.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/mobile/mobile-purchases-feast-google-process-subs-CODE-lambda",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/mobile/mobile-purchases-feast-google-process-subs-CODE-lambda/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "mobilepurchasesfeastgoogleprocesssubsCODElambdaServiceRoleDefaultPolicy6C37C6FB",
        "Roles": [
          {
            "Ref": "mobilepurchasesfeastgoogleprocesssubsCODElambdaServiceRoleFA723C8D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "mobilepurchasesfeastgoogleprocesssubsCODElambdaServiceRoleFA723C8D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "mobile-purchases-feast-google-process-subs-CODE-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "mobilepurchasesfeastgoogleprocesssubsCODElambdamobilepurchasesfeastgoogleprocesssubsCODElambdaSnsExistingIncomingEventsTopic13F26053": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "mobilepurchasesfeastgoogleprocesssubsCODElambdaA1DA9075",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "Region": {
          "Ref": "AWS::Region",
        },
        "TopicArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:sns:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::GetAtt": [
                  "mobilepurchasesfeastgoogleprocesssubsCODEsnsqueue5470C729",
                  "QueueName",
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "mobilepurchasesfeastgoogleprocesssubsCODEsnsqueue5470C729": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "MessageRetentionPeriod": 1209600,
        "QueueName": "mobile-purchases-feast-google-process-subs-CODE-queue",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`The FeastAndroidSubscriptions stack matches the snapshot 2`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuSnsLambdaExperimental",
    ],
    "gu:cdk:version": "TEST",
  },
  "Outputs": {
    "mobilepurchasesfeastgoogleprocesssubsPRODlambdaTopicName06FD04C6": {
      "Value": {
        "Fn::GetAtt": [
          "mobilepurchasesfeastgoogleprocesssubsPRODsnsqueue40111C71",
          "QueueName",
        ],
      },
    },
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "mobilepurchasesfeastgoogleprocesssubsPRODlambda6B998497": {
      "DependsOn": [
        "mobilepurchasesfeastgoogleprocesssubsPRODlambdaServiceRoleDefaultPolicy03AE322B",
        "mobilepurchasesfeastgoogleprocesssubsPRODlambdaServiceRole8B2391E9",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "mobile/PROD/mobile-purchases-feast-google-process-subs-PROD-lambda/mobile-purchases-feast-google-process-subs.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "mobile-purchases-feast-google-process-subs-PROD-lambda",
            "STACK": "mobile",
            "STAGE": "PROD",
          },
        },
        "FunctionName": "mobile-purchases-feast-google-process-subs-PROD-lambda",
        "Handler": "index.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "mobilepurchasesfeastgoogleprocesssubsPRODlambdaServiceRole8B2391E9",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "mobile-purchases-feast-google-process-subs-PROD-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "mobilepurchasesfeastgoogleprocesssubsPRODlambdaAllowInvokefeastsubscriptionsPRODmobilepurchasesfeastgoogleprocesssubsPRODlambdaSnsExistingIncomingEventsTopic5FDCAAB31664DE2C": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "mobilepurchasesfeastgoogleprocesssubsPRODlambda6B998497",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:sns:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::GetAtt": [
                  "mobilepurchasesfeastgoogleprocesssubsPRODsnsqueue40111C71",
                  "QueueName",
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "mobilepurchasesfeastgoogleprocesssubsPRODlambdaServiceRole8B2391E9": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "mobile-purchases-feast-google-process-subs-PROD-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "mobilepurchasesfeastgoogleprocesssubsPRODlambdaServiceRoleDefaultPolicy03AE322B": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/mobile/PROD/mobile-purchases-feast-google-process-subs-PROD-lambda/mobile-purchases-feast-google-process-subs.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/mobile/mobile-purchases-feast-google-process-subs-PROD-lambda",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/mobile/mobile-purchases-feast-google-process-subs-PROD-lambda/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "mobilepurchasesfeastgoogleprocesssubsPRODlambdaServiceRoleDefaultPolicy03AE322B",
        "Roles": [
          {
            "Ref": "mobilepurchasesfeastgoogleprocesssubsPRODlambdaServiceRole8B2391E9",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "mobilepurchasesfeastgoogleprocesssubsPRODlambdamobilepurchasesfeastgoogleprocesssubsPRODlambdaSnsExistingIncomingEventsTopic26530584": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "mobilepurchasesfeastgoogleprocesssubsPRODlambda6B998497",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "Region": {
          "Ref": "AWS::Region",
        },
        "TopicArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:sns:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::GetAtt": [
                  "mobilepurchasesfeastgoogleprocesssubsPRODsnsqueue40111C71",
                  "QueueName",
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "mobilepurchasesfeastgoogleprocesssubsPRODsnsqueue40111C71": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "MessageRetentionPeriod": 1209600,
        "QueueName": "mobile-purchases-feast-google-process-subs-PROD-queue",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
